crypto/x509/test-vectors/generate.sh

```diff
@@ -0,0 +1,366 @@
+#!/bin/bash
+
+set -x
+set -u
+set -e
+
+
+# Some variables we'll use later
+FIFTYYEARS="$(expr 365 \* 50 + 50 / 4)"
+BASEDIR="$(dirname $0)"
+
+OUTPUT_BASE_DIR="${OUT_DIR}/openssl"
+OPENSSL_CNF="${OUT_DIR}/openssl/openssl.cnf"
+
+OK_ROOT_DIR="${OUTPUT_BASE_DIR}/ok_root"
+OK_PENULTIMATE_DIR="${OUTPUT_BASE_DIR}/ok_penultimate"
+OK_INTERMEDIATE1_DIR="${OUTPUT_BASE_DIR}/ok_intermediate1"
+OK_INTERMEDIATE2_DIR="${OUTPUT_BASE_DIR}/ok_intermediate2"
+OK_INTERMEDIATE3_DIR="${OUTPUT_BASE_DIR}/ok_intermediate3"
+OK_INTERMEDIATE4_DIR="${OUTPUT_BASE_DIR}/ok_intermediate4"
+OK_INTERMEDIATE5_DIR="${OUTPUT_BASE_DIR}/ok_intermediate5"
+OK_INTERMEDIATE6_DIR="${OUTPUT_BASE_DIR}/ok_intermediate6"
+OK_INTERMEDIATE7_DIR="${OUTPUT_BASE_DIR}/ok_intermediate7"
+OK_PENULTIMATE8_DIR="${OUTPUT_BASE_DIR}/ok_penultimate8"
+
+
+# Initialize a directory to contain a certificate authority
+function init_ca_dir() {
+	DIRNAME="$1"
+	mkdir -p "${DIRNAME}"
+	mkdir -p "${DIRNAME}/certs"
+	mkdir -p "${DIRNAME}/crl"
+	mkdir -p "${DIRNAME}/newcerts"
+	mkdir -p "${DIRNAME}/private"
+	mkdir -p "${DIRNAME}/req"
+
+	chmod 700 "${DIRNAME}/private"
+	touch "${DIRNAME}/index.txt"
+	echo '1000' > "${DIRNAME}/serial"
+	echo '1000' > "${DIRNAME}/crlnumber"
+}
+
+
+# Create an intermediate CA
+function make_intermediate_ca() {
+	ID=$1
+	PARENT="ok_root"
+	if [[ 1 -lt $ID ]]; then
+		PARENT="ok_intermediate$(expr $ID - 1)"
+	fi
+
+	DIR="OK_INTERMEDIATE${ID}_DIR"
+
+	openssl genpkey \
+		-algorithm rsa \
+		-pkeyopt rsa_keygen_bits:2048 \
+		-outform PEM \
+		-out "${!DIR}/private/ca.key"
```

Comment:
Out of curiosity, whats the `!` for?

Reply:
That's bash's variable-variables:

```bash
FOO="stuff"
BAR="FOO"
echo "${!BAR}"
```

will print "stuff"